lisa/tests/staging/sched_android.py

```diff
@@ -201,4 +201,90 @@ class UnboostedFreq(SchedTuneFreqBoost):
     """
     boost = 0
 
+
+class SchedTuneTaskPlacement(SchedTuneTest, abc.ABC):
+    """
+    Abstract class for testing SchedTune's task-placement biasing feature
+
+    Run a tiny rtapp tasks in a cgroup with prefer-idle set, and checks that
+    it runs most of the time on CPUs with enough capacity to meet its boost
+    request.
+    """
+    ftrace_conf = FtraceConf({
+        "events" : ["sched_switch", "sched_wakeup"]
+    }, __qualname__)
+
+    prefer_idle = True
+
+    @classmethod
+    def check_from_target(cls, target):
+        # XXX: better way to do that ?
+        if not target.file_exists('/sys/kernel/debug/tracing/available_events'):
+            raise CannotCreateError("Ftrace not enabled")
+
+        # XXX: better way to do that ?
+        for m in ['cgroups', 'sched']:
+            if not target.has(m):
+                raise CannotCreateError("'{}' devlib module not enabled".format(m))
+
+        for c in ['schedtune']:
+            if c not in target.cgroups.controllers:
+                raise CannotCreateError("'{}' cgroup controller not available".format(c))
+
+    @classmethod
+    def get_rtapp_profile(cls, plat_info):
+        cpu = plat_info['capacity-classes'][-1][0]
+        rtapp_profile = {}
+        rtapp_profile['rta_stune'] = Periodic(
+            duty_cycle_pct = 1,
+            duration_s = 3,
+            period_ms = 16,
+        )
+
+        return rtapp_profile
+
+    @requires_events('sched_wakeup')
+    def test_task_placement(self, bad_cpu_margin_pct=10) -> ResultBundle:
+        """
+        Test that the task placement satisfied the boost requirement
+
+        Check that top-app tasks spend no more than ```bad_cpu_margin_pct```
+        of their time on CPUs that don't have enough capacity to serve their
+        boost.
+        """
+
+        task = list(self.rtapp_profile.keys())[0]
```

Comment:
could we add `assert len(self.rtapp_profile) == 1` ? That would make sure any unexpected change to number of tasks is not silently ignored by that code.